### PR TITLE
Adding expvar metrics to capture the current number of context in the aggregator

### PR DESCRIPTION
**Pull request by hush-hush**
### What does this PR do?

Adding expvar metrics to capture the current number of context in the aggregator

### Motivation

We want to be able to track the number of context in the aggregator (especially for DogStatsD) to troubleshoot tagging issues that might increase the number of context for a metrics.

### CODE REVIEW

File: pkg/metrics/context_metrics.go [hunk 1]
```diff
@@ -80,3 +80,8 @@ func (m ContextMetrics) Flush(timestamp float64) ([]*Serie, map[ckey.ContextKey]
 
 	return series, errors
 }
+
+// Length returns the number of contexts currently stored
+func (m ContextMetrics) Lenght() int {
```
Comment (olivielpeau) [inline pkg/metrics/context_metrics.go:85]:
```suggestion
func (m ContextMetrics) Length() int {
```

File: pkg/aggregator/aggregator.go [hunk 2]
```diff
@@ -107,13 +107,19 @@ var (
...
+	aggregatorDogstatsdContexts                = expvar.Int{}
+	aggregatorDogstatsdBucketsContexts         = expvar.Int{}
 
 	tlmFlush = telemetry.NewCounter("aggregator", "flush",
 		[]string{"data_type", "state"}, "Number of metrics/service checks/events flushed")
 	tlmProcessed = telemetry.NewCounter("aggregator", "processed",
 		[]string{"data_type"}, "Amount of metrics/services_checks/events processed by the aggregator")
 	tlmHostnameUpdate = telemetry.NewCounter("aggregator", "hostname_update",
 		nil, "Count of hostname update")
+	tlmContexts = telemetry.NewCounter("aggregator", "contexts",
```
Comment (olivielpeau) [inline pkg/aggregator/aggregator.go:119]: should be a gauge

```suggestion
	tlmContexts = telemetry.NewGauge("aggregator", "contexts",
```

File: pkg/aggregator/aggregator.go [hunk 3]
```diff
@@ -107,13 +107,19 @@ var (
...
+		nil, "Count the number of contexts in the aggregator")
```
Comment (olivielpeau) [inline pkg/aggregator/aggregator.go:120]: you need to specify the tag key here (the value will be `dogstatsd`), so maybe:

```suggestion
		[]string{"aggregator"}, "Count the number of contexts in the aggregator")
```

File: pkg/aggregator/context_resolver.go [hunk 4]
```diff
@@ -65,6 +65,15 @@ func (cr *ContextResolver) trackContext(metricSampleContext metrics.MetricSample
 	return contextKey
 }
 
+func (cr *ContextResolver) get(key ckey.ContextKey) (*Context, bool) {
```
Comment (olivielpeau) [inline pkg/aggregator/context_resolver.go:68]: does this have any overhead? 🤔

File: pkg/aggregator/time_sampler.go [hunk 5]
```diff
@@ -188,14 +188,23 @@ func (s *TimeSampler) flush(timestamp float64) (metrics.Series, metrics.SketchSe
...
 	s.lastCutOffTime = cutoffTime
 
+	aggregatorDogstatsdContexts.Set(int64(s.contextResolver.length()))
+	tlmContexts.Set(float64(s.contextResolver.length()), "dogstatsd")
+
+	nbContextsInBuckets := float64(0)
+	for _, contextMetrics := range s.metricsByTimestamp {
+		nbContextsInBuckets += contextMetrics.Length()
+	}
+	tlmBucketsContexts.Set(nbContextsInBuckets)
```
Comment (olivielpeau) [inline pkg/aggregator/time_sampler.go:198]:
```suggestion
	tlmBucketsContexts.Add(nbContextsInBuckets)
```

File: pkg/aggregator/time_sampler.go [hunk 6]
```diff
@@ -188,14 +188,23 @@ func (s *TimeSampler) flush(timestamp float64) (metrics.Series, metrics.SketchSe
...
+	aggregatorDogstatsdBucketsContexts.Set(int64(nbContextsInBuckets))
```
Comment (olivielpeau) [inline pkg/aggregator/time_sampler.go:199]:
```suggestion
	aggregatorDogstatsdBucketsContexts.Add(int64(nbContextsInBuckets))
```

File: pkg/aggregator/context_resolver.go (hunk 4 above)
Comment (hush-hush) [reply pkg/aggregator/context_resolver.go:68]: Doubtful, It's just a function call (which might be optimized).

File: pkg/aggregator/aggregator.go [hunk 7]
```diff
@@ -107,13 +107,16 @@ var (
...
 	aggregatorOrchestratorMetadataErrors       = expvar.Int{}
+	aggregatorDogstatsdContexts                = expvar.Int{}
 
 	tlmFlush = telemetry.NewCounter("aggregator", "flush",
 		[]string{"data_type", "state"}, "Number of metrics/service checks/events flushed")
 	tlmProcessed = telemetry.NewCounter("aggregator", "processed",
 		[]string{"data_type"}, "Amount of metrics/services_checks/events processed by the aggregator")
 	tlmHostnameUpdate = telemetry.NewCounter("aggregator", "hostname_update",
 		nil, "Count of hostname update")
+	tlmContexts = telemetry.NewGauge("aggregator", "contexts",
```
Comment (olivielpeau) [inline pkg/aggregator/aggregator.go:118]: to be explicit:
```suggestion
	tlmDogstatsdContexts = telemetry.NewGauge("aggregator", "dogstatsd_contexts",
```

File: pkg/aggregator/aggregator.go [hunk 8]
```diff
@@ -107,13 +107,16 @@ var (
...
+		nil, "Count the number of contexts in the aggregator")
```
Comment (olivielpeau) [inline pkg/aggregator/aggregator.go:119]:
```suggestion
		nil, "Count the number of dogstatsd contexts in the aggregator")
```

File: pkg/aggregator/time_sampler.go [hunk 9]
```diff
@@ -188,14 +188,16 @@ func (s *TimeSampler) flush(timestamp float64) (metrics.Series, metrics.SketchSe
 	s.contextResolver.expireContexts(timestamp - defaultExpiry)
 	s.lastCutOffTime = cutoffTime
 
+	aggregatorDogstatsdContexts.Set(int64(s.contextResolver.length()))
+	tlmContexts.Set(float64(s.contextResolver.length()))
```
Comment (olivielpeau) [inline pkg/aggregator/time_sampler.go:192]:
```suggestion
	tlmDogstatsdContexts.Set(float64(s.contextResolver.length()))
```

**Review by olivielpeau — APPROVED** [review]
This metric will certainly be helpful.

Just a small renaming suggestion on the telemetry var name, other than that LGTM 👍

**Pull request merged**